tsinfer/inference.py

```diff
@@ -490,6 +607,111 @@ def match_samples(
     return ts
 
 
+def map_additional_mutations(
+    sample_data,
+    tables,
+    *,
+    sample_id_map=None,
+    sites_exist=True,
+    check_consistency=True,
+    progress_monitor=False,
+):
+    """
+    Add mutations into tree sequence tables parsimoneously, using the tskit
+    map_mutations method, such that at site that exist in sample_data but not
+    in the tables, the generated variation corresponds to the distribution seen in
+    the sample_data file. This is done by iterating over the trees.
+
+    Correspondance between the samples in the sample_data file and the ids in the
+    node table is specified by the sample_id_map. If only a subset of sample ids
+    are provided, then only the variation for those samples will be used to map.
+
+    Sites in ``sample_data`` that already exist in the site table are left untouched.
+
+    Sample 0 in sample_data should correspond to node 0 in the input tree sequence,
+    sample 1 to node 1, and so on. If ``consistency_checks`` is True, the samples are
+    checked to see if they belong to the same set of individuals and populations, as
+    assessed by the individual and population metadata, and checks are performed to
+    ensure that the final tree sequence generates the exact pattern of variation
+    contained in the sample_data file.
+    """
+    if sample_id_map is None:
+        sample_id_map = {i: i for i in range(sample_data.num_samples)}
+    if sample_data.sequence_length != tables.sequence_length:
+        raise ValueError(
+            f"The sample_data and tables must have the same sequence length"
+        )
+    ts = tables.tree_sequence()
+    if check_consistency:
+        verify_samples(sample_data, ts, sample_id_map=sample_id_map)
+    trees = ts.trees()
+    tree = next(trees)
+    sample_indexes = np.array(list(sample_id_map.keys()), dtype=int)
+    positions = sample_data.sites_position[:]
+    new_sd_sites = np.where(np.isin(positions, tables.sites.position) == 0)[0]
+    # Create new sites and add the mutations
+    pm = progress_monitor.get("ms_parsimony_mutations", len(new_sd_sites))
+    for variant in sample_data.variants(ids=new_sd_sites):
+        site = variant.site
+        pos = site.position
+        anc_state = site.ancestral_state
+        anc_value = 0  # sample_data files always have 0 as the ancestral allele idx
+        G = variant.genotypes[sample_indexes]
+        if np.all(np.logical_or(G == tskit.MISSING_DATA, G == anc_value)):
```

Comment:
It took a while to parse this - a comment would have helped. So, if the site is just a mixture of missing data of the ancestral state, or the ancestral state, we don't bother doing any inference and impute everything to be ancestral. But, wouldn't Fitch do this anyway? This seems like unnecessary complexity to me, which is (a) unrelated to the purpose of the PR and (b) probably a premature optimisation. Is this corner case actually costing us time? Do we gain anything from identifying it? I would prefer to remove it, and get the real functionality in, and follow it up later if it's needed.

Reply:
We can't use Fitch if all the sites are missing (we previously simply skipped the site if so). I thought that we might as well skip it if there was any mix of missing or ancestral sites. But I could simply restrict it to all-missing sites, as before.

It seemed more informative to mark these sites as not having used the fitch-inference algorithm at all (since we haven't), hence the `INFERENCE_NONE` idea. I think it would be misleading to mark an entirely missing site as INFERENCE_FITCH, to be honest.

Reply:
Ah, OK, thanks, that makes more sense. A comment to that effect would help.

Reply:
So yes, INFERENCE_NONE makes sense now.